cdap-ui/app/cdap/components/PreviewData/Table.tsx

```diff
@@ -0,0 +1,146 @@
+import TableHead from '@material-ui/core/TableHead';
+import TableRow from '@material-ui/core/TableRow';
+import Paper from '@material-ui/core/Paper';
+import TableCell from '@material-ui/core/TableCell';
+import withStyles, { WithStyles } from '@material-ui/core/styles/withStyles';
+import isEmpty from 'lodash/isEmpty';
+import { PREVIEW_STATUS } from 'services/PreviewStatus';
+import Heading, { HeadingTypes } from 'components/Heading';
+import ThemeWrapper from 'components/ThemeWrapper';
+import T from 'i18n-react';
+
+const I18N_PREFIX = 'features.PreviewData.Table';
+
+const CustomTableCell = withStyles((theme) => ({
+  head: {
+    backgroundColor: theme.palette.grey['300'],
+    color: theme.palette.common.white,
+    padding: 10,
+    fontSize: 14,
+  },
+  body: {
+    padding: 10,
+    fontSize: 14,
+  },
+}))(TableCell);
+
+export const messageTextStyle = {
+  fontSize: '1.3rem !important',
+  margin: '10px 0',
+};
+const styles = (theme) => ({
+  root: {
+    width: '100%',
+    display: 'inline-block',
+    height: 'auto',
+    marginTop: theme.spacing(1),
+  },
+  table: {
+    width: '100%',
+  },
+  row: {
+    height: 40,
+    '&:nth-of-type(odd)': {
+      backgroundColor: theme.palette.grey['600'],
+    },
+  },
+  messageText: messageTextStyle,
+});
+
+interface IDataTableProps extends WithStyles<typeof styles> {
+  headers?: string[];
+  records?: any[];
+  isInput?: boolean;
+  previewStatus?: string;
+  isCondition?: boolean;
+}
+
+const DataTableView: React.FC<IDataTableProps> = ({
+  classes,
+  headers,
+  records,
+  isInput,
+  previewStatus,
+  isCondition,
+}) => {
+  const getStatusMsg = () => {
+    let msg;
+    const recordType = isInput ? 'Input' : 'Output';
+    if (isCondition) {
+      msg = T.translate(`${I18N_PREFIX}.previewNotSupported`);
+    } else if (previewStatus === PREVIEW_STATUS.RUNNING || previewStatus === PREVIEW_STATUS.INIT) {
+      // preview is still running but there's no data yet
+      msg = T.translate(`${I18N_PREFIX}.previewRunning`, { recordType });
+    } else {
+      // not running preview but there is no preview data
+      msg = T.translate(`${I18N_PREFIX}.noPreviewRunning`, { recordType });
+    }
+    return msg;
+  };
+
+  if (isEmpty(records) || isCondition) {
+    return (
+      <div>
+        <Heading type={HeadingTypes.h3} label={getStatusMsg()} className={classes.messageText} />
+      </div>
+    );
+  }
+
+  return (
+    <Paper className={classes.root}>
+      <Table>
+        <TableHead>
+          <TableRow className={classes.row}>
+            {headers.map((fieldName, i) => {
+              return <CustomTableCell key={`header-cell-${i}`}>{fieldName}</CustomTableCell>;
+            })}
+          </TableRow>
+        </TableHead>
+        <TableBody>
+          {records.map((record, j) => {
+            return (
+              <TableRow className={classes.row} key={`tr-${j}`}>
+                {headers.map((fieldName, k) => {
```

Comment:
Is it possible to have very wide tables? E.g. hundreds or thousands of columns? If so, does this handle them well?

Reply:
Answered in a meeting. There will be feature work in the future to address large schemas.